replit_river/client.py

```diff
@@ -160,15 +161,27 @@ def _trace_procedure(
     service_name: str,
     procedure_name: str,
 ) -> Generator[Span, None, None]:
-    with tracer.start_span(
+    span = tracer.start_span(
         f"river.client.{procedure_type}.{service_name}.{procedure_name}",
         kind=SpanKind.CLIENT,
-    ) as span:
-        try:
-            yield span
-        except RiverException as e:
-            _record_river_error(span, RiverError(code=e.code, message=e.message))
-            raise e
+    )
+    set_status = False
+    try:
+        yield span
+    except RiverException as e:
+        span.record_exception(e)
+        _record_river_error(span, RiverError(code=e.code, message=e.message))
+        set_status = True
+        raise e
+    except (Exception, CancelledError) as e:
+        span.record_exception(e)
+        span.set_status(StatusCode.ERROR, f"{type(e).__name__}: {e}")
+        set_status = True
+        raise e
+    finally:
+        if not set_status:
```

Comment:
maybe add a comment that the `except` chain will _explicitly_ not catch any `BaseException`s (except `CancelledError`. this is peak [enclave](https://www.youtube.com/watch?v=PrJSdYCkoWk) territory) because those are kinda sorta reserved for Python to do control flow.

Reply:
Adjusted things to specifically catch and raise `GeneratorExit` to make things more clear.